fix: undo operation on signature pad

plxity opened the pull request:
fixes: #864

plxity commented:
@Mythie @dguyen Please review the changes.

adithyaakrishna reviewed (approved):
LGTM 🌮

plxity commented:
@Mythie Bumping this up 😅

catalinpit commented:
I'm on the `main` branch, and the undo functionality works properly. Am I missing something?

https://github.com/documenso/documenso/assets/25515812/b1a94c93-39fa-452c-a08b-6fd7f590a088

plxity commented:
@catalinpit Please watch the video I have attached and the steps to reproduce.

catalinpit commented:
> @catalinpit Please watch the video I have attached and the steps to reproduce.

Yeah, I've seen those. I thought it happens in all scenarios, but the issue actually occurs when you already have a signature and try to update it.

![1](https://github.com/documenso/documenso/assets/25515812/337c122a-65f4-4f8b-8e5b-a9608fb79929)

Thanks for the heads up! Will test the changes and review the PR.

catalinpit commented:
Ok, I just tested it. It works fine on the profile page, but for some reason doesn't work when signing a document. I attached a video.

Does it work for you?

https://github.com/documenso/documenso/assets/25515812/4bbb2142-e5f9-4696-bc8f-b4a575a94f72

plxity commented:
Yup, it seems to be working for me. Can you restart server or clear browser cache and try again?

catalinpit commented:
> Yup, it seems to be working for me. Can you restart server or clear browser cache and try again?

I restarted the server, cleared browser cache, and even ran `npm run reset:hard`, and it still doesn't work unfortunately...

plxity commented:
@catalinpit My bad, I have fixed it. In `useEffect` deps value, there is no need to add `defaultValue`. Because we want that effect to only run once when the component loads, I've fixed it and have tested the changes at all the places. Please review it again.

plxity commented on packages/ui/primitives/signature-pad/signature-pad.tsx:
```diff
@@ -183,8 +184,11 @@ export const SignaturePad = ({
     // Clear the canvas
     if ($el.current) {
       const ctx = $el.current.getContext('2d');
+      const { width, height } = $el.current;
```
Fixed it, I think we can add eslint error for an unused variable😅

catalinpit commented on packages/ui/primitives/signature-pad/signature-pad.tsx:
```diff
@@ -183,8 +184,11 @@ export const SignaturePad = ({
     // Clear the canvas
     if ($el.current) {
```
Another thing I observed is that when you undo the first change to an already existing signature, the signature is not updated. See the video to better understand what I mean.

https://github.com/documenso/documenso/assets/25515812/281b776d-8366-4116-8bb9-0df5d6f1cfdb

I tried calling the `onChange` function after clearing the canvas to update the signature URL and it seems to work.

```js
if ($el.current) {
      const ctx = $el.current.getContext('2d');
      const { width, height } = $el.current;
      ctx?.clearRect(0, 0, width, height);
      if (typeof defaultValue === 'string' && defaultImageRef.current) {
        ctx?.putImageData(defaultImageRef.current, 0, 0);
      }
      newLines.forEach((line) => {
        const pathData = new Path2D(getSvgPathFromStroke(getStroke(line, perfectFreehandOptions)));
        ctx?.fill(pathData);
      });

      /*
         here
      */
      onChange?.($el.current.toDataURL());
    }
```

Please double check and see if it works and/or if there is a better way of doing it.

Here's how it works with the new addition.

https://github.com/documenso/documenso/assets/25515812/10a95739-e0b5-4889-894b-5d57df96c136

plxity replied:
@catalinpit Thanks for pointing out. This makes sense, I've added an onChange handler. This is a good method to work with it :)

plxity replied:
@catalinpit Let me know if the changes are fine :)

catalinpit reviewed (dismissed):
thanks for the work!

Mythie commented:
Great work 😄
